### PR TITLE
improvement: cast wallet import / --account option

### PR DESCRIPTION
**Pull request by rplusq**
### Intro
Most of the motivation comes from these issues:
[Issue#1869](https://github.com/foundry-rs/foundry/issues/1869)
[Issue#3818](https://github.com/foundry-rs/foundry/issues/3818)

Big shoutout to @PatrickAlphaC for the proposed API

### Summary

If you want to use your own private keys, you are going to either: expose them on the terminal or type them every time through `--interactive`. A great alternative would be to use keystores, but these are limited to random ones created through `cast wallet new PATH`. 

It would also be good to be able to set some env vars and then forget about that, while also maintaining a high security standard.

#### cast wallet import

Encrypt private keys through JSON keystores into a default directory (`~/.foundry/keystores`) that can later be accessed through their assigned names when using `cast send`, `forge create` or `forge script`.

```
cast wallet import --interactive
```
Which then prompts the user for a password, and encrypts it locally on the default directory.

#### cast wallet list
Have a command to list stored keystores in the default directory
```
cast wallet list
```

#### Extraction of Wallet's Raw Opts

I had to extract the raw options from the `opts/wallet:Wallet` struct. This is cause I wanted to reduce the amount of code duplication, so now I could use that in the use of `cast wallet import`, which really only cares about the handling of mnemonics and private keys.

#### Adding of --account option for MultiWallet and Wallet

In order to leverage the imported keystores by name, I had to create a new option: `--account` or `ETH_KEYSTORE_ACCOUNT`. This enables to simplify the following command:

```
forge script scripts/<YOUR_SCRIPT> --keystore ~/.foundry/keystotres/<account_name>.json --broadcast
```
into the smaller / simpler:

```
forge script scripts/<YOUR_SCRIPT> --account <account_name> --broadcast
```

### CODE REVIEW

File: cli/src/cmd/cast/wallet/mod.rs [hunk 1]
```diff
@@ -170,6 +189,71 @@ impl WalletSubcommands {
                     }
                 }
             }
+            WalletSubcommands::Import { account_name, keystore_dir, raw_wallet_options } => {
+                // Set up keystore directory
+                let dir = if let Some(path) = keystore_dir {
+                    path
+                } else {
+                    let home_dir = dirs::home_dir().unwrap();
+                    home_dir.join(".foundry").join("keystores").to_string_lossy().to_string()
```
Comment (mattsse) [inline cli/src/cmd/cast/wallet/mod.rs:198]: this should be available via the `Config` functions

File: cli/src/cmd/cast/wallet/mod.rs [hunk 2]
```diff
@@ -170,6 +189,71 @@ impl WalletSubcommands {
...
+                )?;
+
+                let private_key = wallet.signer().to_bytes();
+                let password= rpassword::prompt_password("Enter password: ")?;
+
+                let mut rng = thread_rng();
+                eth_keystore::encrypt_key(&dir, &mut rng, &private_key, &password, Some(&account_name))?;
+            },
+            WalletSubcommands::List { } => {
+                let default_keystore_dir = dirs::home_dir().unwrap().join(".foundry").join("keystores");
```
Comment (mattsse) [inline cli/src/cmd/cast/wallet/mod.rs:227]: same here

File: cli/src/cmd/cast/wallet/mod.rs [hunk 3]
```diff
@@ -170,6 +189,71 @@ impl WalletSubcommands {
...
+                // check if keystore directory exists
+                if !default_keystore_dir.exists() {
+                    eyre::bail!("Keystore directory does not exist at {}", default_keystore_dir.display());
+                }
+                // List all files in keystore directory
+                let keystore_files = std::fs::read_dir(default_keystore_dir)?
+                .filter_map(|entry| {
+                    let entry = entry.unwrap();
```
Comment (mattsse) [inline cli/src/cmd/cast/wallet/mod.rs:235]: please no unwrap here

File: cli/src/opts/wallet/mod.rs [hunk 4]
```diff
@@ -137,34 +156,52 @@ pub struct Wallet {
...
-        Ok(if let Some(ref private_key) = self.private_key {
+        Ok(if let Some(ref private_key) = self.raw.private_key {
             Some(self.get_from_private_key(private_key)?)
         } else {
             None
         })
     }
 
     pub fn keystore(&self) -> Result<Option<LocalWallet>> {
+        let default_keystore_dir = dirs::home_dir()
```
Comment (mattsse) [inline cli/src/opts/wallet/mod.rs:182]: same

File: cli/src/opts/wallet/multi_wallet.rs [hunk 5]
```diff
@@ -282,7 +296,32 @@ impl MultiWallet {
     ///
     /// Returns `Ok(None)` if no keystore provided.
     pub fn keystores(&self) -> Result<Option<Vec<LocalWallet>>> {
-        if let Some(keystore_paths) = &self.keystore_paths {
+        let default_keystore_dir = dirs::home_dir()
+            .ok_or_else(|| eyre::eyre!("Failed to get home directory"))?
+            .join(".foundry")
+            .join("keystores");
```
Comment (mattsse) [inline cli/src/opts/wallet/multi_wallet.rs:302]: same

File: cli/src/opts/wallet/multi_wallet.rs [hunk 6]
```diff
@@ -292,7 +331,7 @@ impl MultiWallet {
                 self.keystore_password_files.clone().unwrap_or_default().into_iter();
 
             for path in keystore_paths {
-                wallets.push(self.get_from_keystore(Some(path), passwords_iter.next().as_ref(), password_files_iter.next().as_ref())?.wrap_err("Keystore paths do not have the same length as provided passwords or password files.")?);
+                wallets.push(self.get_from_keystore(Some(&path), passwords_iter.next().as_ref(), password_files_iter.next().as_ref())?.wrap_err("Keystore paths do not have the same length as provided passwords or password files.")?);
```
Comment (mattsse) [inline cli/src/opts/wallet/multi_wallet.rs:334]: please do this in two steps, the large push( argument makes this harder to read

**Review by mattsse — CHANGES_REQUESTED** [review]
I like this feature,

haven't tested this yet, but left some suggestions.

the config struct has helpers for getting the .foundry dir, we can even add a new function that returns the keystore dir

File: cli/src/cmd/cast/wallet/mod.rs (hunk 3 above)
Comment (rplusq) [reply cli/src/cmd/cast/wallet/mod.rs:235]: done

File: cli/src/opts/wallet/mod.rs (hunk 4 above)
Comment (rplusq) [reply cli/src/opts/wallet/mod.rs:182]: Done

File: cli/src/opts/wallet/multi_wallet.rs (hunk 6 above)
Comment (rplusq) [reply cli/src/opts/wallet/multi_wallet.rs:334]: Done

**Comment by rplusq** [comment]
Thanks for the feedback, now addressed @mattsse.

File: cli/src/cmd/cast/wallet/mod.rs [hunk 7]
```diff
@@ -170,6 +190,84 @@ impl WalletSubcommands {
...
+
+                let mut rng = thread_rng();
+                eth_keystore::encrypt_key(
+                    &dir,
+                    &mut rng,
+                    private_key,
+                    &password,
+                    Some(&account_name),
+                )?;
+            }
```
Comment (Evalir) [inline cli/src/cmd/cast/wallet/mod.rs:234]: I think adding a message that the keystore was saved successfully would be nice! ideally in green

File: cli/src/opts/wallet/mod.rs [hunk 8]
```diff
@@ -27,70 +28,67 @@ pub use multi_wallet::*;
...
-        long,
-        short,
-        help_heading = "Wallet options - raw",
-        value_name = "ADDRESS",
-        env = "ETH_FROM"
-    )]
-    pub from: Option<Address>,
-
+#[clap(next_help_heading = "Wallet options - raw", about = None, long_about = None)]
+pub struct RawWallet {
```
Comment (Evalir) [inline cli/src/opts/wallet/mod.rs:37]: Let's add a bit of docs on what this is (essentially a wrapper for the raw data for `Wallet`)

**Review by Evalir — CHANGES_REQUESTED** [review]
looks good apart from some nits here and there. need to test this feature though—especially to make sure that hw wallets still work properly and such.

File: cli/src/opts/wallet/mod.rs (hunk 8 above)
Comment (rplusq) [reply cli/src/opts/wallet/mod.rs:38]: Done

File: cli/src/cmd/cast/wallet/mod.rs [hunk 9]
```diff
@@ -170,6 +190,84 @@ impl WalletSubcommands {
...
+                        .filter_map(|entry| match entry {
+                            Ok(entry) => {
+                                let path = entry.path();
+                                if path.is_file() && path.extension().is_none() {
+                                    Some(Ok(path))
+                                } else {
+                                    None
+                                }
+                            }
+                            Err(e) => Some(Err(eyre::Report::from(e))),
```
[the remark this answers is not in the document]
Comment (rplusq) [reply cli/src/cmd/cast/wallet/mod.rs:253]: Updated. Is it good like this?

File: cli/src/cmd/cast/wallet/mod.rs (hunk 7 above)
Comment (rplusq) [reply cli/src/cmd/cast/wallet/mod.rs:234]: Using eyre color, or is there a better alternative? Seems like this clutters autocompletion by recommending all colors for everything that could be printed

**Comment by rplusq** [comment]
@Evalir addressed your feedback but left some follow-up questions

File: cli/src/cmd/cast/wallet/mod.rs (hunk 7 above)
Comment (Evalir) [reply cli/src/cmd/cast/wallet/mod.rs:243]: hey! we have usually use `yansi::Paint` for this—there's a few examples in the codebase, feel free to check them out!

**Review by Evalir — CHANGES_REQUESTED** [review]
looking good bar nits—wanna loop in @mds1 for testing

**Comment by PatrickAlphaC** [comment]
OMG LETS GOOOOO RAFAEL!!!!!

File: cli/src/cmd/cast/wallet/mod.rs (hunk 7 above)
Comment (rplusq) [reply cli/src/cmd/cast/wallet/mod.rs:243]: Thanks, done now!

**Review by mattsse — APPROVED** [review]
lgtm, pending @Evalir

**Review by Evalir — APPROVED** [review]
Sick. Let's get this in today!

**Comment by Osneue** [comment]
Good job!

I find here through Partick's solidity course. @PatrickAlphaC

I try to test this new feature.

I import my private key by typing
```
cast wallet import gan-wall --interactive
```

and I get the output
```
Enter private key:
Enter password:
`gan-wall` keystore was saved successfully. Address: 0xdcc0ed94411272c1eb99067f34dfb0aefc511709
```

The address "0xdcc0ed94411272c1eb99067f34dfb0aefc511709" listed above is the same as the address in Ganache corresponding to private key I input. So I think the saved keystore has recorded the address corresponding to the private key.
It all works well up to now.

And then I try to deploy my smart contract using script by typing
```
forge script script/DeploySimpleStorage.s.sol --rpc-url $GANACHE_RPC_URL --broadcast --account gan-wall
```
But I get an error
```
[⠒] Compiling...
No files changed, compilation skipped
Script ran successfully.

== Return ==
0: contract SimpleStorage 0x34A1D3fff3958843C43aD80F30b94c510645C316

EIP-3855 is not supported in one or more of the RPCs used.
Unsupported Chain IDs: 1337.
Contracts deployed with a Solidity version equal or higher than 0.8.20 might not work properly.
For more information, please see https://eips.ethereum.org/EIPS/eip-3855

## Setting up (1) EVMs.

==========================

Chain 1337

Estimated gas price: 4.380813908 gwei

Estimated total gas used for script: 464097

Estimated amount required: 0.002033122592261076 ETH

==========================

###
Finding wallets for all the necessary addresses...
Enter keystore password:

Transactions saved to: /home/sjz/solidity-course/patrick/foundry-simple-storage-f23/broadcast/DeploySimpleStorage.s.sol/1337/run-latest.json

Sensitive values saved to: /home/sjz/solidity-course/patrick/foundry-simple-storage-f23/cache/DeploySimpleStorage.s.sol/1337/run-latest.json

Error:

You seem to be using Foundry's default sender. Be sure to set your own --sender.
No associated wallet for addresses: {0x1804c8ab1f12e6bbf3894d4083f33e07309d1f38}. Unlocked wallets: [0xdcc0ed94411272c1eb99067f34dfb0aefc511709]
```

Obviously it is caused by the wrong account address, so I retry it by typing
```
forge script script/DeploySimpleStorage.s.sol --rpc-url $GANACHE_RPC_URL --broadcast --account gan-wall --sender 0xdCC0eD94411272c1eB99067F34dFb0AEfc511709
```
Finally, everything goes well.

And my question is @rplusq

**Since I have already specified --account, why do I need to specify --sender?**

**Why does the --account option not automatically select the corresponding address?**

I am brand new to this field, please forgive me if my question is stupid.

And please correct me if there is anything wrong with my formatting.@PatrickAlphaC

**Pull request merged**